src/main/java/mat/server/service/FhirValidationReportService.java

```diff
@@ -95,31 +101,79 @@ private String generateValidationReport(ConversionResultDto conversionResultDto,
         paramsMap.put("matVersion", currentMatVersion);
 
         if (conversionResultDto != null && measure != null) {
-            //Measure FHIR validation errors
-            List<FhirValidationResult> measureFhirValidationErrors = conversionResultDto.
-                    getMeasureConversionResults().getMeasureFhirValidationResults();
-
-            // Library FHIR validation errors
-            List<LibraryConversionResults> libraryConversionResults = conversionResultDto.getLibraryConversionResults();
-            List<FhirValidationResult> libraryFhirValidationErrors = libraryConversionResults.stream().flatMap(i -> i.getLibraryFhirValidationResults().stream()).collect(Collectors.toList());
-
-            // CQL conversion errors
-            List<CqlConversionError> cqlConversionErrors = libraryConversionResults.stream().map(i -> i.getCqlConversionResult()).flatMap(i -> i.getCqlConversionErrors().stream()).collect(Collectors.toList());
-
-            // ValueSet FHIR validation errors
-            List<FhirValidationResult> valueSetFhirValidationErrors = conversionResultDto.getValueSetConversionResults().stream().flatMap(i -> i.getValueSetFhirValidationResults().stream()).collect(Collectors.toList());
+            if (!("SUCCESS").equals(conversionResultDto.getOutcome().name())) {
+
+
+                if(!conversionResultDto.getErrorReason().isEmpty()) {
+                    paramsMap.put("ErrorReason", conversionResultDto.getErrorReason());
+                }
+                // ValueSet FHIR validation errors
+                List<FhirValidationResult> valueSetFhirValidationErrors = new ArrayList<>();
+                if (CollectionUtils.isNotEmpty(conversionResultDto.getValueSetConversionResults())) {
+                    for (ValueSetConversionResults valueSetConversionResults : conversionResultDto.getValueSetConversionResults()) {
+                        if (Boolean.FALSE.equals(valueSetConversionResults.getSuccess())) {
+                            valueSetFhirValidationErrors = valueSetConversionResults.getValueSetFhirValidationResults();
+                        }
+                    }
+                }
+
+                // Library FHIR validation errors
+                List<FhirValidationResult> libraryFhirValidationErrors = new ArrayList<>();
+                Set<CqlConversionError> qdmCqlConversionErrors = new HashSet<>();
+                Set<MatCqlConversionException> qdmMatCqlConversionErrors = null;
+                Set<CqlConversionError> fhirCqlConversionErrors = null;
+                Set<MatCqlConversionException> fhirMatCqlConversionErrors = null;
+                if (CollectionUtils.isNotEmpty(conversionResultDto.getLibraryConversionResults())) {
+                    for (LibraryConversionResults results : conversionResultDto.getLibraryConversionResults()) {
```

Comment:
Does codacy complain about cyclomatic complexity where?
The number of nested blocks is quite hight. You might want to consider refactoring that.

Reply:
Not really, It did one I cleared it in the last commit.

Reply:
Maybe Codacy is not that strict as SonarLint.
SonarLint would be red :)